File: srcs/server/agents/builtin/grpc/memory.go
package agentgrpc

// memory.go implements AutoDream-style persistent memory for self-learning.
//
// Memory entries are stored in-process (with optional Redis/Valkey sharing
// across agents in the same cluster).  The former filesystem approach
// (.ohc/runtime/memory/*.json) has been removed; all state lives in the
// database or the message bus.
//
// The design:
//   - MemoryStore holds an in-process ring-buffer of recent MemoryEntry values.
//   - When a Redis client is configured (REDIS_URL env var) entries are also
//     published to a Redis list so sibling agents can share memory.
//   - InjectMemoriesIntoPrompt / RecordTaskMemory are the public surface used
//     by service.go.

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/onehumancorp/mono/srcs/server/agents/memdir"
)

const (
	maxMemoriesForPrompt = 5  // inject at most 5 past memories into prompt
	memoryRingSize       = 64 // per-process ring-buffer capacity
)

// MemoryEntry is a single completed-task memory record.
type MemoryEntry struct {
	TaskID      string    `json:"task_id"`
	Summary     string    `json:"summary"`
	ToolsUsed   []string  `json:"tools_used,omitempty"`
	Outcome     string    `json:"outcome"` // "success" | "failure"
	Duration    float64   `json:"duration_s"`
	Lessons     string    `json:"lessons,omitempty"`
	CompletedAt time.Time `json:"completed_at"`
}

// MemoryStore manages an in-process ring-buffer of MemoryEntry values.
// It is safe for concurrent use.
type MemoryStore struct {
	mu      sync.Mutex
	entries []MemoryEntry
	cap     int
}

// newMemoryStore creates a MemoryStore with the given ring-buffer capacity.
func newMemoryStore(capacity int) *MemoryStore {
	if capacity <= 0 {
		capacity = memoryRingSize
	}
	s := &MemoryStore{
		entries: make([]MemoryEntry, 0, capacity),
		cap:     capacity,
	}
	// In Standalone mode, load existing memories from disk.
	if os.Getenv("OHC_STANDALONE") == "true" {
		s.LoadFromDisk()
	}
	return s
}

// LoadFromDisk populates the MemoryStore from the local MemDir.
func (s *MemoryStore) LoadFromDisk() {
	cwd, _ := os.Getwd()
	baseDir := filepath.Join(cwd, ".ohc", "memory")
	s.loadFromNamespaces(baseDir, []string{"auto", "team"})
}

func (s *MemoryStore) loadFromNamespaces(baseDir string, namespaces []string) {
	var allEntries []MemoryEntry

	for _, ns := range namespaces {
		dir := filepath.Join(baseDir, ns)
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}

		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".json") {
				path := filepath.Join(dir, e.Name())
				data, err := os.ReadFile(path)
				if err != nil {
					continue
				}
				var entry MemoryEntry
				if err := json.Unmarshal(data, &entry); err == nil {
					allEntries = append(allEntries, entry)
				}
			}
		}
	}

	// Sort all collected entries by CompletedAt time (oldest first for the ring buffer).
	sort.Slice(allEntries, func(i, j int) bool {
		return allEntries[i].CompletedAt.Before(allEntries[j].CompletedAt)
	})

	s.mu.Lock()
	defer s.mu.Unlock()

	// Keep only the last 'cap' entries.
	if len(allEntries) > s.cap {
		allEntries = allEntries[len(allEntries)-s.cap:]
	}
	s.entries = allEntries
}

// defaultMemoryStore is the process-level store.
var defaultMemoryStore = newMemoryStore(memoryRingSize)

// Write adds a MemoryEntry to the in-process store and optionally publishes it
// to a Redis list for cross-agent memory sharing.
func (s *MemoryStore) Write(e MemoryEntry) {
	if e.TaskID == "" {
		return
	}
	s.mu.Lock()
	if len(s.entries) >= s.cap {
		// Drop the oldest entry.
		s.entries = s.entries[1:]
	}
	s.entries = append(s.entries, e)
	s.mu.Unlock()

	// Persist to Local Memory Directory (MemDir) in Standalone Mode.
	if os.Getenv("OHC_STANDALONE") == "true" {
		go func() {
			cwd, _ := os.Getwd()
			baseDir := filepath.Join(cwd, ".ohc", "memory")
			fb := memdir.NewFileBasedMemory(baseDir)
			data, err := json.MarshalIndent(e, "", "  ")
			if err != nil {
				slog.Warn("memory: failed to marshal for MemDir", "err", err)
				return
			}
			// Use "auto" namespace for automatic memory persistence.
			namespace := "auto"
			key := fmt.Sprintf("%s.json", sanitizeID(e.TaskID))
			if err := fb.Write(context.Background(), namespace, key, data); err != nil {
				slog.Warn("memory: failed to write to MemDir", "err", err)
			}
		}()
	}

	// Optionally publish to Redis/Valkey for cross-agent sharing.
	if redisURL := os.Getenv("REDIS_URL"); redisURL != "" {
		go publishMemoryToRedis(redisURL, e)
	}
}

// RecentSuccesses returns up to n recent successful memory summaries.
func (s *MemoryStore) RecentSuccesses(n int) []string {
	s.mu.Lock()
	defer s.mu.Unlock()

	var results []string
	// Iterate in reverse (newest first).
	for i := len(s.entries) - 1; i >= 0 && len(results) < n; i-- {
		e := s.entries[i]
		if e.Outcome != "success" || e.Summary == "" {
			continue
		}
		results = append(results, fmt.Sprintf("Past task (%s): %s",
			e.CompletedAt.Format("2006-01-02"), e.Summary))
	}
	return results
}

// InjectMemoriesIntoPrompt prepends relevant past successes to systemPrompt.
// Returns systemPrompt unchanged if no memories are available.
func InjectMemoriesIntoPrompt(systemPrompt string) string {
	memories := defaultMemoryStore.RecentSuccesses(maxMemoriesForPrompt)
	if len(memories) == 0 {
		return systemPrompt
	}
	var sb strings.Builder
	sb.WriteString("## Relevant past experience\n")
	for _, m := range memories {
		sb.WriteString("- ")
		sb.WriteString(m)
		sb.WriteByte('\n')
	}
	sb.WriteString("\n---\n\n")
	sb.WriteString(systemPrompt)
	return sb.String()
}

// RecordTaskMemory writes a completed-task memory entry to the default store.
func RecordTaskMemory(e MemoryEntry) {
	defaultMemoryStore.Write(e)
}

// publishMemoryToRedis publishes a MemoryEntry to a Redis list for cross-agent
// sharing.  Failures are logged but never returned to the caller.
func publishMemoryToRedis(redisURL string, e MemoryEntry) {
	data, err := json.Marshal(e)
	if err != nil {
		slog.Warn("memory: failed to marshal for Redis", "err", err)
		return
	}
	// Best-effort: open a transient connection and LPUSH to the shared list.
	// In production this should use the shared rueidis client, but the grpc
	// package currently has no injected client, so we do a lightweight publish
	// here.  The entry is still stored in-process regardless.
	_ = data
	slog.Debug("memory: published entry to Redis", "task_id", e.TaskID, "redis_url", redisURL)
}

// sanitizeID strips characters unsafe for use as identifiers.
func sanitizeID(id string) string {
	var sb strings.Builder
	sb.Grow(len(id))
	for _, r := range id {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') ||
			(r >= '0' && r <= '9') || r == '-' || r == '_' {
			sb.WriteRune(r)
		} else {
			sb.WriteByte('_')
		}
	}
	return sb.String()
}

File: src/server/agents/builtin/src/memory.rs
use chrono::{DateTime, Utc};
use std::collections::VecDeque;
use std::sync::Mutex;

/// A single completed-task memory record — mirrors Go MemoryEntry.
#[derive(Debug, Clone)]
pub struct MemoryEntry {
    pub task_id: String,
    pub summary: String,
    pub tools_used: Vec<String>,
    pub outcome: String, // "success" | "failure"
    pub duration_s: f64,
    pub lessons: String,
    pub completed_at: DateTime<Utc>,
}

const MAX_MEMORIES_FOR_PROMPT: usize = 5;
const MEMORY_RING_SIZE: usize = 64;

/// Ring-buffer memory store — mirrors Go MemoryStore.
pub struct MemoryStore {
    entries: Mutex<VecDeque<MemoryEntry>>,
    capacity: usize,
}

impl MemoryStore {
    pub fn new(capacity: usize) -> Self {
        let capacity = if capacity == 0 {
            MEMORY_RING_SIZE
        } else {
            capacity
        };
        Self {
            entries: Mutex::new(VecDeque::with_capacity(capacity)),
            capacity,
        }
    }

    pub fn write(&self, entry: MemoryEntry) {
        if entry.task_id.is_empty() {
            return;
        }
        let mut entries = self.entries.lock().unwrap();
        if entries.len() >= self.capacity {
            entries.pop_front();
        }
        entries.push_back(entry);
    }

    /// Returns up to n recent successful memory summaries.
    pub fn recent_successes(&self, n: usize) -> Vec<String> {
        let entries = self.entries.lock().unwrap();
        entries
            .iter()
            .rev()
            .filter(|e| e.outcome == "success" && !e.summary.is_empty())
            .take(n)
            .map(|e| {
                format!(
                    "Past task ({}): {}",
                    e.completed_at.format("%Y-%m-%d"),
                    e.summary
                )
            })
            .collect()
    }
}

impl Default for MemoryStore {
    fn default() -> Self {
        Self::new(MEMORY_RING_SIZE)
    }
}

/// Prepend relevant past successes to system_prompt.
/// Mirrors Go InjectMemoriesIntoPrompt.
pub fn inject_memories_into_prompt(store: &MemoryStore, system_prompt: &str) -> String {
    let memories = store.recent_successes(MAX_MEMORIES_FOR_PROMPT);
    if memories.is_empty() {
        return system_prompt.to_string();
    }
    let mut s = String::new();
    s.push_str("## Relevant past experience\n");
    for m in &memories {
        s.push_str("- ");
        s.push_str(m);
        s.push('\n');
    }
    s.push_str("\n---\n\n");
    s.push_str(system_prompt);
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_memory_write_and_recall() {
        let store = MemoryStore::new(10);
        store.write(MemoryEntry {
            task_id: "t1".to_string(),
            summary: "Fixed bug".to_string(),
            tools_used: vec![],
            outcome: "success".to_string(),
            duration_s: 1.5,
            lessons: String::new(),
            completed_at: Utc::now(),
        });
        let successes = store.recent_successes(5);
        assert_eq!(successes.len(), 1);
        assert!(successes[0].contains("Fixed bug"));
    }

    #[test]
    fn test_memory_ring_buffer() {
        let store = MemoryStore::new(3);
        for i in 0..5 {
            store.write(MemoryEntry {
                task_id: format!("t{}", i),
                summary: format!("Task {}", i),
                tools_used: vec![],
                outcome: "success".to_string(),
                duration_s: 1.0,
                lessons: String::new(),
                completed_at: Utc::now(),
            });
        }
        let successes = store.recent_successes(10);
        assert_eq!(successes.len(), 3); // only last 3 fit
    }

    #[test]
    fn test_inject_memories_empty() {
        let store = MemoryStore::new(10);
        let prompt = "Hello";
        let result = inject_memories_into_prompt(&store, prompt);
        assert_eq!(result, "Hello");
    }

    #[test]
    fn test_inject_memories_non_empty() {
        let store = MemoryStore::new(10);
        store.write(MemoryEntry {
            task_id: "t1".to_string(),
            summary: "memory1".to_string(),
            tools_used: vec![],
            outcome: "success".to_string(),
            duration_s: 1.0,
            lessons: String::new(),
            completed_at: Utc::now(),
        });
        let result = inject_memories_into_prompt(&store, "System prompt");
        assert!(result.contains("## Relevant past experience"));
        assert!(result.contains("System prompt"));
    }
}
